### ISSUE TITLE
t.news.database should use t.m.smtp.sendmail and the email package

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4528]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)| @cyli reported|
|-|-|
|Trac ID|trac#4528|
|Type|task|
|Created|2010-06-29 21:24:20Z|
|Branch|https://github.com/twisted/twisted/tree/news-sendmail-4528|

Currently it uses t.m.smtp.sendEmail, which has been deprecated since mid-2003 and should be removed.  (See [#4038](https://github.com/twisted/twisted/issues/4038))

<details><summary>Searchable metadata</summary>

```
trac-id__4528 4528
type__task task
reporter__cyli cyli
priority__normal normal
milestone__ 
branch__branches_news_sendmail_4528 branches/news-sendmail-4528
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__news news
keywords__ 
time__1277846660000000 1277846660000000
changetime__1279054521000000 1279054521000000
version__None None
owner__ 
cc__therve
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_2"></a>|@exarkun commented|
|-|-|

(In [29547]) Branching to 'news-sendmail-4528'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_3"></a>|@exarkun removed owner|
|-|-|

Switched to `sendmail` in the branch.

[build results](http://buildbot.twistedmatrix.com/boxes-supported?branch=/branches/news-sendmail-4528)

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_4"></a>|@cyli set owner to @exarkun|
|-|-|

Pyflakes reports no problems, tests pass, no whitespace, so on a superficial level it looks great.

However, it looks like this branch changes the functionality/API of t.news.database slightly:
 1. Previously, the sender was always 'twisted@[something]' ([something] being socket.gethostname()) for both PickleStorage and NewsShelf.  Now, the sender can be None (and is in fact, by default, None).
 1. Previously, NewsShelf actually had the method `notifyModerator`, which took one moderator.  Now it provides `notifyModerators` which takes multiple moderators.
 1. The subject also seems to be provided now, as well as Content-Type, which is probably just better.

There's also the fact that I don't see how PickleStorage.notifyModerators even worked in the past, since it makes a call to smtp.sendEmail without the mailhost argument.  Now notifyModerators actually seems to work, which I guess isn't really a change in functionality but a bugfix - the fact that there are now tests for this is also a good thing.

It looks like PickleStorage.notifyModerators is only called from PickleStorage.postRequest, and NewsShelf.notifyModerators is only called from NewsShelf.postRequest, so maybe all of these concerns just don't matter.  I don't see a docstring in twisted.news that or twisted.news.database that says that this is all internal code though, and not a publicly-exposed API - perhaps one should be added?  If it is supposed to be a publicly facing API, perhaps some comment should be added to the news fragment regarding the change.

I'm not really familiar with who uses this, or how it's used, so I'm not sure these concerns should stop the branch from being merged.  Since you were the maintainer in the first place and are one of the core developers, if you decide that the slight content and functionality changes are fine, then I'd say looks good to merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_5"></a>|@exarkun removed owner|
|-|-|

Thanks for the review.

> Previously, the sender was always 'twisted@[something]' ([something] being socket.gethostname()) for both PickleStorage and NewsShelf. Now, the sender can be None (and is in fact, by default, None).

I've put back a default so that these won't try to send mail from `None`. (r29583)

> Previously, NewsShelf actually had the method notifyModerator, which took one moderator. Now it provides notifyModerators which takes multiple moderators.

I added back `notifyModerator` to `NewsShelf` as a wrapper around `notifyModerators`. (r29582)

> The subject also seems to be provided now, as well as Content-Type, which is probably just better.

Yay. :)  I'm not going to worry about the contents of the email from a backwards compatibility stand-point.

> There's also the fact that I don't see how PickleStorage.notifyModerators even worked in the past

Indeed.  I don't think it worked before either.

> It looks like PickleStorage.notifyModerators is only called from PickleStorage.postRequest, and NewsShelf.notifyModerators is only called from NewsShelf.postRequest, so maybe all of these concerns just don't matter.

Yea... They were probably intended to be implementation details, but the fact that they aren't named to indicate they are private (leading _) means they are public.  So even though it sucks, we must treat them as public APIs.  If someone cares enough, then we should probably eventually deprecate them and put private versions in their place.  It doesn't seem like a major issue to worry about now, though.

> If it is supposed to be a publicly facing API, perhaps some comment should be added to the news fragment regarding the change.

I'll mention it, since it is public, even though it isn't supposed to be. ;)  (r29584)

Thanks for the great review.  All of the points you brought up made sense.

I also added some docstrings for the mail-related parameters (r29585), too.

**Comment by twisted-trac** [comment]
|[<img alt="therve's avatar" src="https://avatars.githubusercontent.com/u/159137?s=50" width="50" height="50">](https://github.com/therve)<a name="note_6"></a>|@therve set owner to @exarkun|
|-|-|

 * ModerationTestsMixin.sendmail has more than 80 chars, and there is a small indentation weirdness after the "append".
 * A blank line is missing before getApprovedMessage.
 * `+        msg['From'] = self._sender`: this should use "sender" instead, as self._sender is a new attribute.

+1!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_7"></a>|@exarkun commented|
|-|-|

(In [29813]) <80 cols

refs [#4528](https://github.com/twisted/twisted/issues/4528)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_8"></a>|@exarkun commented|
|-|-|

(In [29814]) whitespace before getApprovedMessage

refs [#4528](https://github.com/twisted/twisted/issues/4528)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_9"></a>|@exarkun commented|
|-|-|

(In [29815]) Make sure that the "From" header gets a good value as well

refs [#4528](https://github.com/twisted/twisted/issues/4528)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_10"></a>|@exarkun set status to `closed`|
|-|-|

(In [29816]) Merge news-sendmail-4528

Author: exarkun
Reviewer: cyli, therve
Fixes: [#4528](https://github.com/twisted/twisted/issues/4528)

Switch the moderation email generation code in Twisted News from the
deprecated `sendEmail` API to `twisted.mail.smtp.sendmail` and the
`email` package.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
